core/src/main/scala/com/salesforce/op/stages/impl/insights/RecordInsightsLOCO.scala

```diff
@@ -86,26 +124,102 @@ class RecordInsightsLOCO[T <: Model[T]]
 
     val k = $(topK)
     var i = 0
-    while (i < filledSize) {
-      val (oldInd, oldVal) = featureArray(i)
-      featureArray.update(i, (oldInd, 0))
-      val score = modelApply(labelDummy, OPVector(Vectors.sparse(featureSize, featureArray))).score
-      val diffs = baseScore.zip(score).map{ case (b, s) => b - s }
-      val max = diffs.maxBy(math.abs)
-      maxHeap.enqueue((i, max, diffs))
-      if (i >= k) maxHeap.dequeue()
-      featureArray.update(i, (oldInd, oldVal))
-      i += 1
+    val top = $(topKStrategy) match {
+      case s if s == TopKStrategy.Abs.entryName || problemType == ProblemType.MultiClassification => {
+        val absoluteMaxHeap = PriorityQueue.empty(AbsScore)
+        var count = 0
+        while (i < filledSize) {
+          val (oldInd, oldVal) = featureArray(i)
+          featureArray.update(i, (oldInd, 0))
+          val score = modelApply(labelDummy, OPVector(Vectors.sparse(featureSize, featureArray))).score
+          val diffs = baseScore.zip(score).map { case (b, s) => b - s }
+          val max = diffs.maxBy(math.abs)
+          if (max != 0) { // Not keeping LOCOs with value 0.0
+            absoluteMaxHeap.enqueue((i, max, diffs))
+            count += 1
+            if (count > k) absoluteMaxHeap.dequeue()
+          }
+          featureArray.update(i, (oldInd, oldVal))
+          i += 1
+        }
+
+        val topAbs = absoluteMaxHeap.dequeueAll
+
+
+        topAbs.sortBy { case (_, v, _) => -math.abs(v) }
+
+      }
+      case s if s == TopKStrategy.PositiveNegative.entryName => {
+        // Heap that will contain the top K positive LOCO values
+        val positiveMaxHeap = PriorityQueue.empty(MinScore)
+        // Heap that will contain the top K negative LOCO values
+        val negativeMaxHeap = PriorityQueue.empty(MaxScore)
+        // for each element of the feature vector != 0.0
+        // Size of positive heap
+        var positiveCount = 0
+        // Size of negative heap
+        var negativeCount = 0
+        for {i <- 0 until filledSize} {
+          val (oldInd, oldVal) = featureArray(i)
+          val diffs = computeDiffs(i, oldInd, featureArray, featureSize, baseScore)
+          val max = if (problemType == ProblemType.Regression) diffs(0) else diffs(1)
```

Comment:
this seems very hacky and error prone - `if (problemType == ProblemType.Regression) diffs(0) else diffs(1)`, perhaps add a helper function?

Reply:
this is where you would use the indexToExamine variable I suggested above